### PR TITLE
A2-3-1: Exclude wide string literals and utf8 string literals

### LINKED ISSUE #311: `A2-3-1`:  Reports `StringLiteral`s incorrectly
### Affected rules
 - `A2-3-1`

### Description

This rule has an exception:
> It is permitted to use other characters inside the text of a wide string and a UTF-8 encoded string literal.

This rule is currently implemented as two queries, one of which covers the StringLiteral case. We should remove this query as it is covered by the exception.

### PR DESCRIPTION
**Pull request by rvermeulen**
## Description

This fixes #311 

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [ ] Queries have been added for the following rules:
     - _rule number here_
 - [x] Queries have been modified for the following rules:
     - cpp/autosar/invalid-character-in-string-literal

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [x] Yes
  - [ ] No

🚨🚨🚨
_**Reviewer:**_ Confirm that format of *shared* queries (not the .qll file, the
.ql file that imports it) is valid by running them within VS Code. 
  - [x] Confirmed

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [x] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### Reviewer 

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### CODE REVIEW

File: change_notes/2024-01-17-fix-reported-fp-for-a2-3-1.md [hunk 1]
```diff
@@ -0,0 +1,2 @@
+`A2-3-1`: ` cpp/autosar/invalid-character-in-string-literal`
+    - Exclude wide string literals and utf8 string literal.
```
Comment (knewbury01) [inline change_notes/2024-01-17-fix-reported-fp-for-a2-3-1.md:2]: the smallest suggestion, do we want to add the references to issue numbers where the fp was reported? it seems we used to do this like [here](https://github.com/github/codeql-coding-standards/blob/main/change_notes/2021-07-15-base-class-slicing.md)

**Review by knewbury01 — APPROVED** [review]
other than that one very small suggestion, looks good to me!

File: cpp/autosar/test/rules/A2-3-1/test.cpp [hunk 2]
```diff
@@ -4,9 +4,15 @@
...
+const wchar_t *euro2 = L"α";     // COMPLIANT
+const char *euro3 = u8"α";       // COMPLIANT
 
 int valid;
 /*
 Invalid character ↦ NON_COMPLIANT
+*/
+
+/*
+Valid character @ in comments COMPLIANT
```
Comment (knewbury01) [inline cpp/autosar/test/rules/A2-3-1/test.cpp:17]: oh wait, what was this addition meant to be? if it was going to address the other exception case, add that , or else rm this

**Review by knewbury01 — CHANGES_REQUESTED** [review]
am I missing seeing the exception char in this testcase? - sorry , missed this on the first review iteration!

File: change_notes/2024-01-17-fix-reported-fp-for-a2-3-1.md (hunk 1 above)
Comment (rvermeulen) [reply change_notes/2024-01-17-fix-reported-fp-for-a2-3-1.md:2]: That is a good suggestion. Let me add that.

File: cpp/autosar/test/rules/A2-3-1/test.cpp (hunk 2 above)
Comment (rvermeulen) [reply cpp/autosar/test/rules/A2-3-1/test.cpp:17]: This is an exception that was not part of our tests yet. Therefore I added it.

Comment (knewbury01) [reply cpp/autosar/test/rules/A2-3-1/test.cpp:17]: aha yep! understood now!

**Review by knewbury01 — APPROVED** [review]
previous comment was just my misunderstanding, all is looking good

**Pull request merged**